aquadoggo/src/materializer/tasks/dependency.rs

```diff
@@ -1,9 +1,293 @@
+/// the store, composing a "reduce" task for this specific document view.
+async fn construct_relation_task(
+    context: &Context,
+    document_view_id: DocumentViewId,
+) -> Result<Option<Task<TaskInput>>, TaskError> {
+    match context
+        .store
+        .get_document_view_by_id(&document_view_id)
+        .await
+        .map_err(|_| TaskError::Critical)?
+    {
+        Some(_) => Ok(None),
+        None => Ok(Some(Task::new(
+            "reduce",
+            TaskInput::new(None, Some(document_view_id)),
+        ))),
+    }
+}
+
+pub async fn dependency_task(context: Context, input: TaskInput) -> TaskResult<TaskInput> {
+    // Here we retrive the document view by document view id.
+    let document_view = match input.document_view_id {
+        Some(view_id) => match context
+            .store
+            .get_document_view_by_id(&view_id)
+            .await
+            .map_err(|_| TaskError::Critical)?
+        {
+            Some(document_view) => Ok(document_view),
+            // If no document view for the id passed into this task could be retrieved then this
+            // document has been deleted or the document view id was invalid. As "dependency" tasks
+            // are only dispatched after a successful "reduce" task, neither `None` case should
+            // happen, so this is a critical error.
+            None => Err(TaskError::Critical),
+        },
+        // We only expect to handle document_view_ids in a dependency task.
+        None => Err(TaskError::Critical),
+    }?;
+
+    let mut next_tasks = Vec::new();
+
+    // First we handle all pinned or unpinned relations defined in this tasks document view.
+    // We can think of these as "child" relations.
+    for (_key, document_view_value) in document_view.fields().iter() {
+        match document_view_value.value() {
+            p2panda_rs::operation::OperationValue::Relation(_relation) => {
+                // This is a relation to a document, if it doesn't exist in the db yet, then that
+                // means we either have no entries for this document, or we are not materialising
+                // it for some reason. We don't want to kick of a "reduce" or "dependency" task in
+                // either of these cases, however in the future we may want to flag it as a "wanted"
+                // schema.
+            }
+            p2panda_rs::operation::OperationValue::RelationList(_relation_list) => {
+                // same as above...
+            }
+            p2panda_rs::operation::OperationValue::PinnedRelation(pinned_relation) => {
+                // These are pinned relations. We may have the operations for these views in the db,
+                // but this view wasn't pinned yet, so hasn't been materialised. To make sure it is
+                // materialised when possible, we dispatch a "reduce" task for any pinned relations
+                // which aren't found.
```

Comment:
Not sure I understand this comment. What is the difference between the pinned and unpinned scenario that makes us act in this case and not in the other one? This could also be nice to read in a docstring for the whole function.

Reply:
Pinned relations point at documents which will also be represented in their latest state (from the perspective of the node). Materializing the "latest" state for a document was done in the `reduce` task (its the first thing which gets dispatched for every new incoming operation), therefore the `dependency` task doesn't need to take care of that anymore.